tools/dotnet-linker/SetupStep.cs

```diff
@@ -27,78 +27,9 @@ public class SetupStep : ConfigurationAwareStep {
 			}
 		}
 
-		void InsertBefore (IStep step, string stepName)
-		{
-			for (int i = 0; i < Steps.Count; i++) {
-				if (Steps [i].GetType ().Name == stepName) {
-					Steps.Insert (i, step);
-					return;
-				}
-			}
-			DumpSteps ();
-			throw new InvalidOperationException ($"Could not insert {step} before {stepName} because {stepName} wasn't found.");
-		}
-
-		void InsertAfter (IStep step, string stepName)
-		{
-			for (int i = 0; i < Steps.Count;) {
-				if (Steps [i++].GetType ().Name == stepName) {
-					Steps.Insert (i, step);
-					return;
-				}
-			}
-			DumpSteps ();
-			throw new InvalidOperationException ($"Could not insert {step} after {stepName} because {stepName} wasn't found.");
-		}
 
 		protected override void TryProcess ()
 		{
-			// Don't use --custom-step to load each step, because this assembly
-			// is loaded into the current process once per --custom-step,
-			// which makes it very difficult to share state between steps.
```

Comment:
It doesn't look like https://github.com/mono/linker/issues/1314 has been fixed yet?

Reply:
That's true. @marek-safar has been encouraging the steps to move to the official API despite this limitation. For xamarin-android we avoided shared state by moving a cache to be built-in to the linker (more context [here](https://github.com/xamarin/xamarin-android/pull/5748#discussion_r606634212)). I didn't notice a case (yet) where this would cause problems for xamarin-macios - are you aware of any?

Reply:
Yeah, we have several mechanisms how to share a state between steps more effectively. The issue can be still fixed but it's probably a low priority now if we don't have valid use case for it

Reply:
> I didn't notice a case (yet) where this would cause problems for xamarin-macios - are you aware of any?

hmm... there's state being shared. Maybe the confusion is that not all _legacy_ steps are present in the dotnet version ?

A, quite important, missing one is `[Core]TypeMapStep`.

It might be better to pause moving customization code to the target file until the port (all features) are complete.

Reply:
> I didn't notice a case (yet) where this would cause problems for xamarin-macios - are you aware of any?

Yes, the problem is that you can't share data using types defined in the custom step assembly. The custom step assembly is loaded once for each --custom-step argument, and they're considered as different assemblies, which means the types in the assemblies have different identifies for each time the assembly is loaded. So using this API:

https://github.com/mono/linker/blob/795106705fedfceb6d1de60999673dfe39dc7acf/src/linker/ref/Linker/Annotations.cs#L34-L35

won't work, because the second --custom-step argument will get back an object whose type is from the first --custom-step argument's assembly load.

Reply:
Thanks!
> Maybe the confusion is that not all legacy steps are present in the dotnet version ?

Yes, I meant to say that the already-ported steps seemed to work without shared state. However I think that is also not true for the reason @rolfbjarne pointed out. I had thought that the static `ConditionalWeakTable<LinkContext, LinkerConfiguration>` on `LinkerConfiguration` was somehow avoiding the problem, but now I realize it is the same since `LinkerConfiguration` is loaded once per assembly so its static members exist on different type identities. I will revert those changes until we fix the linker issue.